ledger/ledgercore/persistedacctdata.go

```diff
@@ -530,61 +527,172 @@ func (e *ExtendedAssetHolding) ConvertToGroups(assets map[basics.AssetIndex]basi
+					exists = true
+				}
+			}
+			if !exists {
+				count := int(e.Groups[i-1].Count + e.Groups[i].Count)
+				crs = append(crs, continuosRange{i - 1, 2, count})
 			}
-			prevAssetIndex = e.Groups[currentGroup].MinAssetIndex
-			prevGroupIndex = currentGroup
 		}
-		e.Groups[currentGroup].groupData.AssetOffsets[i%MaxHoldingGroupSize] = a.aidx - prevAssetIndex
-		e.Groups[currentGroup].groupData.Amounts[i%MaxHoldingGroupSize] = a.holdings.Amount
-		e.Groups[currentGroup].groupData.Frozens[i%MaxHoldingGroupSize] = a.holdings.Frozen
-		prevAssetIndex = a.aidx
+		loaded = append(loaded, i)
+	}
+	if len(loaded) == 0 {
+		return nil, nil
 	}
-	e.Groups[currentGroup].DeltaMaxAssetIndex = uint64(flatten[len(flatten)-1].aidx - e.Groups[currentGroup].MinAssetIndex)
+
+	return
 }
 
-// Loaded return a boolean flag indicated if groups are loaded or not
-func (e ExtendedAssetHolding) Loaded() bool {
-	return e.loaded
+// merge merges groups [start, start+size) and returns keys of deleted group data entries
+func (e *ExtendedAssetHolding) merge(start int, size int, hint int) (deleted []int64) {
+	deleted = make([]int64, 0, hint)
+	// process i and i + 1 groups at once => size-1 iterations
+	i := 0
+	for i < size-1 {
+		li := start + i     // left group index, destination
+		ri := start + i + 1 // right group index, source
+		lg := &e.Groups[li]
+		rg := &e.Groups[ri]
+
+		num := int(MaxHoldingGroupSize - lg.Count)
+		if num == 0 { // group is full, skip
+			i++
+			continue
+		}
+		if num > int(rg.Count) { // source group is shorter than dest capacity, adjust
+			num = int(rg.Count)
+		}
+		groupDelta := rg.MinAssetIndex - (lg.MinAssetIndex + basics.AssetIndex(lg.DeltaMaxAssetIndex))
+		delta := basics.AssetIndex(0)
+		lg.groupData.AssetOffsets = append(lg.groupData.AssetOffsets, rg.groupData.AssetOffsets[0]+groupDelta)
+		for j := 1; j < num; j++ {
+			lg.groupData.AssetOffsets = append(lg.groupData.AssetOffsets, rg.groupData.AssetOffsets[j])
+			delta += rg.groupData.AssetOffsets[j]
+		}
+		lg.DeltaMaxAssetIndex += uint64(delta + groupDelta)
+		lg.groupData.Amounts = append(lg.groupData.Amounts, rg.groupData.Amounts[:num]...)
+		lg.groupData.Frozens = append(lg.groupData.Frozens, rg.groupData.Frozens[:num]...)
+		lg.Count += uint32(num)
+		if num != int(rg.Count) {
+			// src group survived, update it and repeat
+			rg.Count -= uint32(num)
+			rg.groupData.AssetOffsets = rg.groupData.AssetOffsets[num:]
+			delta += rg.groupData.AssetOffsets[0]
+			rg.groupData.AssetOffsets[0] = 0
+			rg.groupData.Amounts = rg.groupData.Amounts[num:]
+			rg.groupData.Frozens = rg.groupData.Frozens[num:]
+			rg.MinAssetIndex += delta
+			rg.DeltaMaxAssetIndex -= uint64(delta)
+			i++
+		} else {
+			// entire src group gone: save the key and delete from Groups
+			deleted = append(deleted, e.Groups[ri].AssetGroupKey)
+			if ri == len(e.Groups) {
+				// last group, cut and exit
+				e.Groups = e.Groups[:len(e.Groups)-1]
+				return
+			}
```

Comment:
this condition is invalid and removed in params-1 PR as part of generalization for params